### PR TITLE
Fix TestServer from blocking on request stream

### PR DESCRIPTION
**Pull request by JamesNK**
Fixes https://github.com/aspnet/AspNetCore/issues/15415

`HttpContent.ReadAsStreamAsync()` will block when called in custom `HttpContent` implementations. Calling `HttpContent.CopyToAsync(stream)` mirrors what `HttpClient` does more closely.

Note, this PR is on 3.1. The bug blocks people from using TestServer with gRPC. Would be nice to unblock before 5.0. @anurse is this your call? Need anymore info?

### CODE REVIEW

**Comment by davidfowl** [comment]
Do it in 5.0 then backport to 3.1?

File: src/Hosting/TestHost/src/ClientHandler.cs
Comment (davidfowl) [inline src/Hosting/TestHost/src/ClientHandler.cs]: What's weird about this logic is that it assumes the body will be re-used again when you do StreamContent.CopyToAsync.

File: src/Hosting/TestHost/test/Utilities.cs
Comment (davidfowl) [inline src/Hosting/TestHost/test/Utilities.cs]: We should delete these in another PR

File: src/Hosting/TestHost/src/ClientHandler.cs
Comment (JamesNK) [reply src/Hosting/TestHost/src/ClientHandler.cs]: Not having it will break anyone who relied on https://github.com/aspnet/AspNetCore/pull/15591/files#diff-5e75529c0ee5684e3ab7c04e5a7f63a9L121

Comment (davidfowl) [reply src/Hosting/TestHost/src/ClientHandler.cs]: To make it a bit more solid it makes sense to use the Stream instead of rewinding it and throwing it away. Something like:

```C#
var requestPipe = new Pipe();

_ = Task.Run(async () =>
{
    if (requestContent is StreamContent)
    {
        var body = await requestContent.ReadAsStreamAsync();
        if (body.CanSeek)
        {
            // This body may have been consumed before, rewind it.
            body.Seek(0, SeekOrigin.Begin);
        }
        
        await body.CopyToAsync(requestPipe.Writer); // This works 😄 
    }
    else
    {
        await requestContent.CopyToAsync(requestPipe.Writer.AsStream());
    }
    
    await requestPipe.Writer.CompleteAsync();
});

req.Body = new AsyncStreamWrapper(requestPipe.Reader.AsStream(), () => contextBuilder.AllowSynchronousIO);
```

File: src/Hosting/TestHost/src/ClientHandler.cs
Comment (davidfowl) [inline src/Hosting/TestHost/src/ClientHandler.cs]: This task should be awaited on the request processing side, here. https://github.com/aspnet/AspNetCore/blob/79ef2f6e472fe9e7d26562cca56cb38a44ae2e0e/src/Hosting/TestHost/src/HttpContextBuilder.cs#L99

It's also where you could complete the reader side of the pipe.

**Comment by Tratcher** [comment]
> `HttpContent.ReadAsStreamAsync()` will block when called in custom `HttpContent` implementations.

Which custom HttpContent implementations? It sounds like the content objects are at fault here. I'm not sure why you're changing the server to accommodate them.

File: src/Hosting/TestHost/src/ClientHandler.cs
Comment (Tratcher) [inline src/Hosting/TestHost/src/ClientHandler.cs]: How do you cancel this loop?

**Comment by Tratcher** [comment]
> Do it in 5.0 then backport to 3.1?

Forward is easier than backwards.

File: src/Hosting/TestHost/src/ClientHandler.cs
Comment (davidfowl) [reply src/Hosting/TestHost/src/ClientHandler.cs]: It's easy to cancel this, the question is when should it be cancelled?

**Comment by davidfowl** [comment]
> Which custom HttpContent implementations? It sounds like the content objects are at fault here. I'm not sure why you're changing the server to accommodate them.

Sorta but not really. It's much more complicated to implement this in each HttpContent. See https://github.com/dotnet/corefx/blob/26d7b6626ab0385a1158e674855f81ab73f628e0/src/System.Net.Http/src/System/Net/Http/HttpContent.cs#L453

I had the same impression at first but it's far easier to make this work in TestServer than it is to fix the HttpContent implementation (just like we auto-rewind your Stream, though I don't, agree with that one).

> Forward is easier than backwards.

But then we can't merge it until it goes through the layers of approval. But whatever works I guess.

File: src/Hosting/TestHost/src/ClientHandler.cs
Comment (JamesNK) [reply src/Hosting/TestHost/src/ClientHandler.cs]: HttpContent.CopyToAsync doesn't support CancellationToken. It is being added in 5.0. I'll add a comment here linking to the issue.

**Comment by JamesNK** [comment]
> Which custom HttpContent implementations?

https://github.com/grpc/grpc-dotnet/blob/9e39edb0a3fc7a625085490bf64c2794870a519e/src/Grpc.Net.Client/Internal/PushStreamContent.cs

> It sounds like the content objects are at fault here. I'm not sure why you're changing the server to accommodate them.

No TestServer is at fault for using `HttpContent.ReadAsStreamAsync` instead of `HttpContent.CopyToAsync`. `HttpContent.ReadAsStreamAsync` is never called on the request content by `HttpClientHandler`.

Should the gRPC content fix itself so ReadAsStreamAsync is supported without blocking? Maybe. But that's a separate issue.

File: src/Hosting/TestHost/src/ClientHandler.cs
Comment (davidfowl) [reply src/Hosting/TestHost/src/ClientHandler.cs]: The workaround is a wrapping stream that does support cancellation

**Comment by JamesNK** [comment]
Related question in HttpClient-side - https://github.com/dotnet/corefx/issues/42160

File: src/Hosting/TestHost/src/ClientHandler.cs
Comment (JamesNK) [reply src/Hosting/TestHost/src/ClientHandler.cs]: Ehhh, do you want to do this now? `ReadAsStreamAsync` that this is replacing doesn't support cancellation either. I'd rather leave this as is and then add a CT when https://github.com/dotnet/corefx/issues/9071 is done in 5.0.

Comment (davidfowl) [reply src/Hosting/TestHost/src/ClientHandler.cs]: Only if we need to cancel that’s why I asked when do we need to cancel.

Comment (Tratcher) [reply src/Hosting/TestHost/src/ClientHandler.cs]: ReadAsStreamAsync wasn't cancellable but usually returned quickly and then you could use cancellation with individual reads.

When would we cancel this loop? When the app didn't read the whole body or aborted the response.

Comment (davidfowl) [reply src/Hosting/TestHost/src/ClientHandler.cs]: Sounds good. Are we assuming a poorly written client? What happens when HttpResponseMessage.Dispose is called? Does this work as expected? What’s broken if we don’t support cancellation? What will hang and in what cases?

Comment (JamesNK) [reply src/Hosting/TestHost/src/ClientHandler.cs]: Using SendAsync's cancellation token:

```cs
var cancelPipeRegistration = cancellationToken.Register(() =>
{
    requestPipe.Writer.Complete(new OperationCanceledException(cancellationToken));
    requestPipe.Reader.Complete(new OperationCanceledException(cancellationToken));
});
```

Comment (davidfowl) [reply src/Hosting/TestHost/src/ClientHandler.cs]: Add the test and the idea isn’t bad but you can call complete while something else is writing to the PipeWriter. CancelPendingFlush is enough. Actually change that to call both CancelPendingRead and CancelPendingFlush and that should be enough

**Comment by Tratcher** [comment]
> No TestServer is at fault for using `HttpContent.ReadAsStreamAsync` instead of `HttpContent.CopyToAsync`. `HttpContent.ReadAsStreamAsync` is never called on the request content by `HttpClientHandler`.
> 
> Should the gRPC content fix itself so ReadAsStreamAsync is supported without blocking? Maybe. But that's a separate issue.

Content that expect to stream really should implement CreateContentReadStreamAsync and then you wouldn't have to change anything in TestServer, no?

File: src/Hosting/TestHost/src/ClientHandler.cs
Comment (JamesNK) [reply src/Hosting/TestHost/src/ClientHandler.cs]: Done

**Comment by davidfowl** [comment]
This change to TestServer is fine

File: src/Hosting/TestHost/test/Utilities.cs
Comment (JamesNK) [reply src/Hosting/TestHost/test/Utilities.cs]: https://github.com/aspnet/AspNetCore/issues/16620

**Comment by JamesNK** [comment]
Also, looking for review approvals // @davidfowl, @Tratcher

**Comment by anurse** [comment]
TestServer is a pretty low-risk area. I'm in favor for 3.1. It's @Pilchie 's final call though.

File: src/Hosting/TestHost/src/ClientHandler.cs
Comment (Tratcher) [inline src/Hosting/TestHost/src/ClientHandler.cs]: One behavior change here is that this initiates client side activity before the app starts reading the request body. I'm not sure what the consequences of that will be in people's tests. In a normal app that would break 100-continue, but in a test app it might be OK because the client doesn't have any automatic behaviors like redirects or auth.

File: src/Hosting/TestHost/src/HttpContextBuilder.cs
Comment (Tratcher) [inline src/Hosting/TestHost/src/HttpContextBuilder.cs]: What about server initiated aborts? Or if the app just completes without completely reading the request body?
https://github.com/aspnet/AspNetCore/blob/0da5fabdbdd462cdd9528398efb6b1ffb9b27f65/src/Hosting/TestHost/src/HttpContextBuilder.cs#L98-L105

File: src/Hosting/TestHost/src/ClientHandler.cs
Comment (Tratcher) [inline src/Hosting/TestHost/src/ClientHandler.cs]: Why does the response body block completion and wait for the request body to drain? What if the server didn't want to read the request body? See my server side abort question.

The app could always drain the request body if it wanted to, but it's not mandatory. The copyRequestTask would also block on buffer space if the app isn't reading it, so this could hang.

**Review by Tratcher — CHANGES_REQUESTED** [review]

File: src/Hosting/TestHost/src/ClientHandler.cs
Comment (davidfowl) [reply src/Hosting/TestHost/src/ClientHandler.cs]: No it doesn't need to be configurable.

File: src/Hosting/TestHost/src/ClientHandler.cs
Comment (davidfowl) [reply src/Hosting/TestHost/src/ClientHandler.cs]: It might be worth fixing this by passing a func to the stream and calling it the first time ReadAsync is called.

File: src/Hosting/TestHost/src/ClientHandler.cs
Comment (davidfowl) [reply src/Hosting/TestHost/src/ClientHandler.cs]: We need to await the task somewhere, there shouldn't be anything doing a fire and forget.

Comment (Tratcher) [reply src/Hosting/TestHost/src/ClientHandler.cs]: Yes, but we also need to guarantee that it completes or gets cancelled. Doing this at the end of the response body doesn't make sense regardless. When the app finishes would be better: https://github.com/aspnet/AspNetCore/blob/0da5fabdbdd462cdd9528398efb6b1ffb9b27f65/src/Hosting/TestHost/src/HttpContextBuilder.cs#L98-L105 is probob

Comment (davidfowl) [reply src/Hosting/TestHost/src/ClientHandler.cs]: Agreed. Actually I initially told @JamesNK to put it there 😄

File: src/Hosting/TestHost/src/ClientHandler.cs
Comment (JamesNK) [reply src/Hosting/TestHost/src/ClientHandler.cs]: > One behavior change here is that this initiates client side activity before the app starts reading the request body.

This isn't a change from how TestServer currently works. If you're passing in `StreamContent` then that stream is read from directly. Any other `HttpContent` implementation was already being eagerly read before the server stated via `ReadAsStreamAsync`.

Comment (Tratcher) [reply src/Hosting/TestHost/src/ClientHandler.cs]: > If you're passing in StreamContent then that stream is read from directly.

But the timing is different. It now starts reading that stream immediately, not when the app starts reading it.

> Any other HttpContent implementation

... that didn't implement CreateContentReadStream.

I'm not strictly worried about this part of the change, I just wanted to call out that the timing change and it may affect tests. All of our tests passed so that implies the effect is limited.

File: src/Hosting/TestHost/src/ClientHandler.cs
Comment (JamesNK) [reply src/Hosting/TestHost/src/ClientHandler.cs]: Done

File: src/Hosting/TestHost/src/HttpContextBuilder.cs
Comment (JamesNK) [reply src/Hosting/TestHost/src/HttpContextBuilder.cs]: I've added a unit test for the server not reading request and server abort.

File: src/Hosting/TestHost/src/ClientHandler.cs
Comment (JamesNK) [reply src/Hosting/TestHost/src/ClientHandler.cs]: Yes you're right. Ok!

**Comment by JamesNK** [comment]
🆙 📅

File: src/Hosting/TestHost/src/HttpContextBuilder.cs
Comment (Tratcher) [inline src/Hosting/TestHost/src/HttpContextBuilder.cs]: You still want to cancel the _requestPipe before waiting for this task.

File: src/Hosting/TestHost/test/TestClientTests.cs
Comment (Tratcher) [inline src/Hosting/TestHost/test/TestClientTests.cs]: This isn't relevant in the TestServer scenario, right?

File: src/Hosting/TestHost/test/TestClientTests.cs
Comment (Tratcher) [inline src/Hosting/TestHost/test/TestClientTests.cs]: Scenario I expect to hang: here you write more data than will fit into the pipe buffer and hit backpressure. The request finishes without reading the request data. The server hangs waiting for the request body to finish.

When the app finishes without draining the request body, the server needs to cancel the body, not drain it.

**Comment by Pilchie** [comment]
What is the scenario for `TestHost` is it (as the name implies) purely for testing, or is it something that customers use on a regular basis in production that we could break?

File: src/Hosting/TestHost/test/TestClientTests.cs
Comment (JamesNK) [reply src/Hosting/TestHost/test/TestClientTests.cs]: Ok. How do I test if the request pipe still has data in it? Something like this?

```cs
if (!_requestPipe.Reader.TryRead(out var result) || !result.IsCompleted)
{
    AbortRequest();
}
else
{
    await _requestPipe.Writer.CompleteAsync();
    await _requestPipe.Reader.CompleteAsync();
}
```

Also how do you cancel the request body? `_requestPipe.Writer.CancelPendingFlush()` doesn't appear to cancel the blocked `_requestPipe.Writer.AsStream().WriteAsync(...)`.

**Comment by anurse** [comment]
> What is the scenario for `TestHost` is it (as the name implies) purely for testing, or is it something that customers use on a regular basis in production that we could break?

It's for testing only. Using it in a production is not supported.

Essentially it's an in-memory test server. You can point it at your `Startup` class and it will start up a fake "app" and give you an `HttpClient` that can make fake "requests" to your app.

**Comment by davidfowl** [comment]
> What is the scenario for TestHost is it (as the name implies) purely for testing, or is it something that customers use on a regular basis in production that we could break?

People use it in Azure functions to bootstrap ASP.NET Core on top of Azure functions.

File: src/Hosting/TestHost/test/TestClientTests.cs
Comment (davidfowl) [reply src/Hosting/TestHost/test/TestClientTests.cs]: I wouldn't do try read. Just drain the thing or cancel it then await the task.

Comment (JamesNK) [reply src/Hosting/TestHost/test/TestClientTests.cs]: It should be canceled if there is blocked content in the unit test because of flow control to propagate the error to the unit test (mirrors what would happen if writing to request when it was completed)

I could make it always abort (aka cancel read and flush). Why is that better?

File: src/Hosting/TestHost/src/HttpContextBuilder.cs
Comment (davidfowl) [inline src/Hosting/TestHost/src/HttpContextBuilder.cs]: Don't do this here, complete the writer on the code doing the writing. Completing the reading should be fine to do here before the task ends.

**Comment by Pilchie** [comment]
Ok, Approved for 3.1.0-Preview3.

File: src/Hosting/TestHost/src/HttpContextBuilder.cs
Comment (JamesNK) [reply src/Hosting/TestHost/src/HttpContextBuilder.cs]: I did it this way because the alternative is callbacks everywhere:
1. After request is complete
2. Request is aborted from server
3. Request is aborted from client

Sure HttpContextBuilder owns the request pipe and task, but using it is optional.

Comment (davidfowl) [reply src/Hosting/TestHost/src/HttpContextBuilder.cs]: Yea I don’t think this layering is super important

**Comment by JamesNK** [comment]
Is there any more feedback/changes required here?

**Review by Tratcher — APPROVED** [review]

**Pull request merged**
